pkgs/development/libraries/gstreamer/good/default.nix

```diff
@@ -143,6 +153,8 @@ stdenv.mkDerivation rec {
     libjack2
   ];
 
+  propagatedBuildInputs = libsoup_3.propagatedUserEnvPackages;
```

Comment:
TBH, I would want to remove `propagatedUserEnvPackages`.

Reply:
I think we should either decide that leaf packages/shells are responsible for pulling glib-networking, or that it should be a hard dependency of libsoup and try to load it internally. Not sure how feasible the latter is though, though I am definitely not a fan of this middle road.

Reply:
i see -- i assumed it was a workaround for the infinite recursion that happens if glib-networking is added the propogatedBuildInputs of libsoup_3.

Reply:
> I think we should either decide that leaf packages/shells are responsible for pulling glib-networking

this seems not ideal as that requires adding glib-networking to the 500-1000 packages that are using gstreamer. if a leaf package forgets to add it there is no hard failure. things work as long as there isn't https or proxies (?) used.

my preference would be to not use it at all nor libsoup and rely on the curl plugin but that plugin doesn't seek correctly and doesn't work on various encoded formats like opus, ogg, webp, etc.

> or that it should be a hard dependency of libsoup and try to load it internally. Not sure how feasible the latter is though, though I am definitely not a fan of this middle road.

putting it in propagatedBuildInputs for libsoup seems better than where it is now, except for the infinite recursion problem. from grepping around in the libsoup and glib source it is not clear to me how it is getting loaded but it seems that glib takes care of scanning and loading the modules in GIO_EXTRA_MODULES. the interface exists for libsoup regardless of if the module is loaded or not, glib just returns an error if libsoup tries to use the functionality and glib-networking is not loaded. perhaps glib provides an interface to explicitly load modules out of some directory rather than relying on GIO_EXTRA_MODULES -- in that case libsoup could request that glib-netwokring modules are loaded at startup.

still would need to solve the infinite recursion problem in the build tho.

Reply:
Ah, it is actually `passthru.propagatedUserEnvPackages` not `propagatedUserEnvPkgs` so that does not do anything at all, opened a [PR to remove it](https://github.com/NixOS/nixpkgs/pull/276813). The latter would cause an infinite recursion since it just adds a reference to the listed derivations to `$out/nix-support/propagated-user-env-packages`, where `nix-env` (and apparently [also `nix shell` now](https://github.com/NixOS/nix/blob/e23983db824f9ea0ecdb6cab3739232b2554f2f5/doc/manual/src/release-notes/rl-2.1.md?plain=1#L35)).

> this seems not ideal as that requires adding glib-networking to the 500-1000 packages that are using gstreamer. if a leaf package forgets to add it there is no hard failure. things work as long as there isn't https or proxies (?) used.

I would expect most packages would not need networking, and if they do, it is an easy fix.

> except for the infinite recursion problem. from grepping around in the libsoup and glib source it is not clear to me how it is getting loaded

If you look use `--show-trace` as recommended by the error, you will get the following chain:

- libsoup-3.4.4
- glib-networking-2.78.0
- libproxy-0.4.18
- networkmanager-1.44.2
- fix-paths.patch
- openconnect-9.12
- stoken-0.93
- gtk+3-3.24.38
- tracker-3.6.0

This cycle should disappear after #272353 is merged.

Reply:
with the cycle removed, this patch to [libsoup/soup-init.c](https://gitlab.gnome.org/GNOME/libsoup/-/blob/2a592b114092b0105f9f27af17225a3b35c288f8/libsoup/soup-init.c) loads glib-networking on init.
```diff
diff --git a/libsoup/soup-init.c b/libsoup/soup-init.c
index 8a33c77..4814c4c 100644
--- a/libsoup/soup-init.c
+++ b/libsoup/soup-init.c
@@ -11,6 +11,7 @@
 
 #include <glib/gi18n-lib.h>
 #include <gmodule.h>
+#include <gio/gio.h>
 #include "gconstructor.h"
 
 #ifdef G_OS_WIN32
@@ -53,6 +54,8 @@ soup_init (void)
 
         if (soup2_is_loaded ())
                 g_error ("libsoup2 symbols detected. Using libsoup2 and libsoup3 in the same process is not supported.");
+
+        g_io_modules_scan_all_in_directory_with_scope("@nixGlibNetworking@", g_io_module_scope_new (G_IO_MODULE_SCOPE_BLOCK_DUPLICATES));
 }
 
 #if defined (G_OS_WIN32)
```

Reply:
> > this seems not ideal as that requires adding glib-networking to the 500-1000 packages that are using gstreamer. if a leaf package forgets to add it there is no hard failure. things work as long as there isn't https or proxies (?) used.
> 
> I would expect most packages would not need networking, and if they do, it is an easy fix.

i'd rather not break things in the first place. this seems like the least desirable option to me. what is wrong with putting glib-networking into the propogatedBuildInputs of libsoup?

Reply:
> i'd rather not break things in the first place. this seems like the least desirable option to me. what is wrong with putting glib-networking into the propogatedBuildInputs of libsoup?

while waiting for an answer to this i'll take a guess. the desire to minimize the closure size is greater than the desire to make it easy to produce a working package when using libsoup.

i sampled 70 leaf packages yesterday that directly included gst-plugins-good. 17% of them were some sort music program with capacity for streaming -- none of them worked though 2 required accounts so they could've worked. some crashed on launch. ones that linked directly with libsoup did not get glib-networking. I fixed one,
https://github.com/NixOS/nixpkgs/pull/276960

the rest i thought would be just as good removed from nixpkgs rather than spending any time on.